File: lib/core/save/checksum.dart
/// SHA-256 checksum over the payload (D4). Pure Dart via package:crypto, so it
/// works on every platform including web.
library;

import 'dart:convert';

import 'package:crypto/crypto.dart';

/// Hex sha256 of the canonical JSON encoding of [payload].
///
/// The payload is encoded with a stable map order so checksums are
/// reproducible across writes: Dart's [JsonEncoder] uses [Map] insertion
/// order, and our models build maps in a fixed order, so two identical saves
/// produce identical checksums.
String checksumOf(Map<String, Object?> payload) {
  final bytes = utf8.encode(jsonEncode(payload));
  return sha256.convert(bytes).toString();
}

/// Constant-time-ish comparison is unnecessary here (integrity check, not
/// auth), but this reads clearer than `!=` and avoids accidental substring
/// mistakes.
bool checksumMatches(Map<String, Object?> payload, String expected) =>
    checksumOf(payload) == expected;

File: packages/core/src/save/checksum.ts
/**
 * SHA-256 checksum over the save payload.
 * Uses the Web Crypto API (available in all modern runtimes).
 */

async function sha256Hex(input: string): Promise<string> {
  const data = new TextEncoder().encode(input);
  const hashBuffer = await crypto.subtle.digest("SHA-256", data);
  const hashArray = new Uint8Array(hashBuffer);
  return Array.from(hashArray)
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

/** Hex sha256 of the canonical JSON encoding of payload. */
export async function checksumOf(
  payload: Record<string, unknown>,
): Promise<string> {
  // Stable JSON encoding: sort keys for reproducibility
  const json = JSON.stringify(payload, Object.keys(payload).sort());
  return sha256Hex(json);
}

/** Check if checksum matches. */
export async function checksumMatches(
  payload: Record<string, unknown>,
  expected: string,
): Promise<boolean> {
  const actual = await checksumOf(payload);
  return actual === expected;
}
